File: plugins/plugin-discord/python/elizaos_plugin_discord/actions/delete_message.py
"""Delete message action for Discord."""

from typing import TYPE_CHECKING

from elizaos_plugin_discord.error import InvalidArgumentError
from elizaos_plugin_discord.types import Snowflake

if TYPE_CHECKING:
    from elizaos_plugin_discord.actions import ActionContext, ActionResult
    from elizaos_plugin_discord.service import DiscordService


class DeleteMessageAction:
    """Action to delete a message from a Discord channel.

    Can delete the bot's own messages unconditionally. For other users'
    messages, the bot needs the Manage Messages permission.
    """

    @property
    def name(self) -> str:
        return "DISCORD_DELETE_MESSAGE"

    @property
    def description(self) -> str:
        return "Delete a message from a Discord channel. Can delete own messages or others' with Manage Messages permission."

    @property
    def similes(self) -> list[str]:
        return [
            "REMOVE_MESSAGE",
            "UNSEND_MESSAGE",
            "DELETE_DISCORD_MESSAGE",
        ]

    async def validate(self, context: "ActionContext") -> bool:
        """Validate that the context is from Discord and has a message_id."""
        source = context.message.get("source")
        if not isinstance(source, str) or source != "discord":
            return False

        try:
            Snowflake(context.channel_id)
        except Exception:
            return False

        # Check that we have a message_id to delete
        state = context.state
        content = context.message.get("content", {})
        if not isinstance(content, dict):
            content = {}

        message_id = ""
        if state:
            message_id = state.get("message_id", "")
        if not message_id:
            message_id = content.get("message_id", "")

        return bool(message_id)

    async def handler(
        self,
        context: "ActionContext",
        service: "DiscordService",
    ) -> "ActionResult":
        """Delete a message from a Discord channel."""
        from elizaos_plugin_discord.actions import ActionResult

        state = context.state or {}
        content = context.message.get("content", {})
        if not isinstance(content, dict):
            content = {}

        # Extract message_id
        message_id = state.get("message_id", "") or content.get("message_id", "")
        if not message_id:
            raise InvalidArgumentError("Missing message_id to delete")

        # Validate message_id as snowflake
        try:
            Snowflake(message_id)
        except Exception:
            raise InvalidArgumentError(f"Invalid message_id: {message_id}")

        channel_id = context.channel_id

        try:
            # Get the channel
            channel = await service._get_text_channel(channel_id)

            # Fetch the message
            target_message = await channel.fetch_message(int(message_id))

            if target_message is None:
                return ActionResult.failure_result(
                    "I couldn't find the message to delete."
                )

            if service._client is None:
                return ActionResult.failure_result(
                    "Discord client is not available."
                )

            bot_user = service._client.user
            is_own_message = (
                bot_user is not None
                and str(target_message.author.id) == str(bot_user.id)
            )

            if not is_own_message:
                # Need Manage Messages permission for other users' messages
                has_perm = await service.has_manage_messages_permission(channel_id)
                if not has_perm:
                    return ActionResult.failure_result(
                        "I don't have permission to delete that message. "
                        "I need the 'Manage Messages' permission to delete "
                        "messages from other users."
                    )

            # Perform the delete
            await target_message.delete()

            return ActionResult.success_result(
                "I've deleted the message.",
                {
                    "message_id": message_id,
                    "channel_id": channel_id,
                },
            )
        except InvalidArgumentError:
            raise
        except Exception as exc:
            return ActionResult.failure_result(f"Failed to delete message: {exc}")

File: plugins/plugin-discord/typescript/actions/deleteMessage.ts
import {
  type Action,
  type ActionExample,
  type ActionResult,
  composePromptFromState,
  type HandlerCallback,
  type HandlerOptions,
  type IAgentRuntime,
  type Memory,
  ModelType,
  parseJSONObjectFromText,
  type State,
} from "@elizaos/core";
import type { Message, TextChannel } from "discord.js";
import { DISCORD_SERVICE_NAME } from "../constants";
import type { DiscordService } from "../service";

const deleteMessageTemplate = `You are helping to extract delete message parameters.

The user wants to delete a Discord message.

Recent conversation:
{{recentMessages}}

Extract the following:
1. messageId: The ID of the message to delete
2. channelRef: The channel where the message is (default: "current")

Respond with a JSON object like:
{
  "messageId": "123456789",
  "channelRef": "current"
}

Only respond with the JSON object, no other text.`;

interface DeleteMessageParams {
  messageId: string;
  channelRef?: string;
}

const deleteMessage: Action = {
  name: "DELETE_MESSAGE",
  similes: [
    "REMOVE_MESSAGE",
    "UNSEND_MESSAGE",
    "DELETE_DISCORD_MESSAGE",
  ],
  description: "Delete a message from a Discord channel",
  
  validate: async (_runtime: IAgentRuntime, message: Memory, _state?: State): Promise<boolean> => {
    return message.content.source === "discord";
  },

  handler: async (
    runtime: IAgentRuntime,
    message: Memory,
    state?: State,
    _options?: HandlerOptions,
    callback?: HandlerCallback
  ): Promise<ActionResult | undefined> => {
    const discordService = (await runtime.getService(DISCORD_SERVICE_NAME)) as DiscordService;

    if (!discordService || !discordService.client) {
      await callback?.({
        text: "Discord service is not available.",
        source: "discord",
      });
      return { success: false, error: "Discord service not available" };
    }

    // Ensure state is available
    const currentState = state ?? (await runtime.composeState(message));
    
    // Use LLM to extract delete parameters
    const prompt = composePromptFromState({
      state: currentState,
      template: deleteMessageTemplate,
    });

    let deleteParams: DeleteMessageParams | null = null;

    for (let i = 0; i < 3; i++) {
      const response = await runtime.useModel(ModelType.TEXT_SMALL, {
        prompt,
      });

      const parsedResponse = parseJSONObjectFromText(response) as unknown as DeleteMessageParams | null;
      if (parsedResponse?.messageId) {
        deleteParams = parsedResponse;
        break;
      }
    }

    if (!deleteParams) {
      await callback?.({
        text: "I couldn't determine which message to delete.",
        source: "discord",
      });
      return { success: false, error: "Failed to extract delete parameters" };
    }

    try {
      // Get the channel
      let channel: TextChannel | null = null;
      
      if (!deleteParams.channelRef || deleteParams.channelRef === "current") {
        const channelId = message.content.channelId as string;
        if (channelId) {
          channel = discordService.client.channels.cache.get(channelId) as TextChannel;
        }
      } else {
        // Try to find channel by name or ID
        channel = discordService.client.channels.cache.find(
          (c) => c.id === deleteParams!.channelRef || 
                 (c.isTextBased() && 'name' in c && c.name === deleteParams!.channelRef)
        ) as TextChannel;
      }

      if (!channel || !channel.isTextBased()) {
        await callback?.({
          text: "I couldn't find the channel with that message.",
          source: "discord",
        });
        return { success: false, error: "Channel not found" };
      }

      // Fetch and delete the message
      const targetMessage = await channel.messages.fetch(deleteParams.messageId) as Message;
      
      if (!targetMessage) {
        await callback?.({
          text: "I couldn't find the message to delete.",
          source: "discord",
        });
        return { success: false, error: "Message not found" };
      }

      // Check if we have permission to delete
      // We can delete our own messages or messages in channels where we have MANAGE_MESSAGES
      const canDelete = 
        targetMessage.author.id === discordService.client.user?.id ||
        (channel.permissionsFor(discordService.client.user!)?.has('ManageMessages') ?? false);

      if (!canDelete) {
        await callback?.({
          text: "I don't have permission to delete that message.",
          source: "discord",
        });
        return { success: false, error: "No permission to delete message" };
      }

      await targetMessage.delete();

      await callback?.({
        text: "I've deleted the message.",
        source: "discord",
      });

      return {
        success: true,
        data: {
          messageId: deleteParams.messageId,
          channelId: channel.id,
        },
      };
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : String(error);
      await callback?.({
        text: `Failed to delete message: ${errorMessage}`,
        source: "discord",
      });
      return { success: false, error: errorMessage };
    }
  },

  examples: [
    [
      {
        name: "{{name1}}",
        content: {
          text: "Delete message 123456789",
        },
      },
      {
        name: "{{agentName}}",
        content: {
          text: "I'll delete that message now.",
          actions: ["DELETE_MESSAGE"],
        },
      },
    ],
    [
      {
        name: "{{name1}}",
        content: {
          text: "Remove that spam message",
        },
      },
      {
        name: "{{agentName}}",
        content: {
          text: "I'll remove that message.",
          actions: ["DELETE_MESSAGE"],
        },
      },
    ],
  ] as ActionExample[][],
};

export default deleteMessage;
